exchange/bitswap/network/ipfs_impl.go

```diff
@@ -38,15 +40,30 @@ type impl struct {
 }
 
 func (bsnet *impl) newStreamToPeer(ctx context.Context, p peer.ID) (inet.Stream, error) {
-
 	// first, make sure we're connected.
 	// if this fails, we cannot connect to given peer.
 	//TODO(jbenet) move this into host.NewStream?
 	if err := bsnet.host.Connect(ctx, peer.PeerInfo{ID: p}); err != nil {
 		return nil, err
 	}
 
-	return bsnet.host.NewStream(ctx, ProtocolBitswap, p)
+	s, err := bsnet.host.NewStream(ctx, ProtocolBitswap, p)
+	if err != nil {
+		return nil, err
+	}
+
+	// TEMP: trigger read handshake to detect protocl mismatch
+	// TODO: once the network has switched over entirely to the versioned
+	// protocol string, deprecate the old one
+	_, err = s.Read(nil)
+	if err != nil {
+		if strings.Contains(err.Error(), "protocol mismatch") {
+			return bsnet.host.NewStream(ctx, ProtocolBitswapOld, p)
+		}
+		return nil, err
+	}
```

Comment:
I'm onboard with the idea of remembering who prefers what, but i'm not sure where that should be stored, how it should be managed, and when it should be cleared. We could use a small lru cache for this i guess, and rely on larger churn to clear out entries of older nodes preferences